File: python/16.py
import re
import sys
from pathlib import Path
from typing import Any, List, Optional, Tuple

sys.path.append(str(Path(__file__).parent.parent.parent))
from Modules.timer import Timer
def part1(data):
    """ 2021 Day 16 Part 1

    >>> part1(['8A004A801A8002F478'])
    16
    >>> part1(['620080001611562C8802118E34'])
    12
    >>> part1(['C0015000016115A2E0802F182340'])
    23
    >>> part1(['A0016C880162017C3686B18A3D4780'])
    31
    """

    packetBinary = bin(int(data[0], 16))[2:]
    packetBinary = '0' * ((8 - (len(packetBinary) % 8)) % 8) + packetBinary
    packet = Packet(packetBinary)
    packet.parse()

    return packet.version + versionSums(packet.subPackets)


def part2(data):
    """ 2021 Day 16 Part 2

    >>> part2(['C200B40A82'])
    3
    >>> part2(['04005AC33890'])
    54
    >>> part2(['880086C3E88112'])
    7
    >>> part2(['CE00C43D881120'])
    9
    >>> part2(['D8005AC2A8F0'])
    1
    >>> part2(['F600BC2D8F'])
    0
    >>> part2(['9C005AC2F8F0'])
    0
    >>> part2(['9C0141080250320F1802104A08'])
    1
    """

    packetBinary = bin(int(data[0], 16))[2:]
    packetBinary = '0' * ((8 - (len(packetBinary) % 8)) % 8) + packetBinary
    packet = Packet(packetBinary)
    packet.parse()

    return packet.eval()


class Packet:
    def __init__(self, packetString):
        self.binary = packetString
        self.parsed = False
        self.subPackets = []
        self.version = 0
        self.typeID = 0
        self.lengthType = 0
        self.value = 0

    def parse(self):
        if self.parsed:
            return

        self.parsed = True
        self.version = int(self.binary[:3], 2)
        self.typeID = int(self.binary[3:6], 2)

        if self.typeID == 4:
            # Literal Packet
            return self.parseLiteral(self.binary[6:])
        else:
            # Operator Packet
            self.subPackets = []
            self.lengthType = int(self.binary[6])

            if self.lengthType == 0:
                # Total Bits
                totalBits = int(self.binary[7:22], 2)
                remainder = self.binary[22:]

                while len(self.binary) - 22 - len(remainder) < totalBits:
                    subPacket = Packet(remainder)
                    self.subPackets.append(subPacket)
                    remainder = subPacket.parse()
                
                return remainder
            else:
                # Number of packets
                numPackets = int(self.binary[7:18], 2)
                remainder = self.binary[18:]
                
                for _ in range(numPackets):
                    subPacket = Packet(remainder)
                    self.subPackets.append(subPacket)
                    remainder = subPacket.parse()

                return remainder        

    def parseLiteral(self, literalString):
        i = 0
        number = ''
        while i < len(literalString):
            number += literalString[i + 1:i + 5]
            i += 5

            if literalString[i - 5] == '0':
                break            

        self.value = int(number, 2)
        return literalString[i:]

    def eval(self):
        if self.typeID == 0:
            # Sum Packet
            total = 0
            for sP in self.subPackets:
                total += sP.eval()

            return total
        if self.typeID == 1:
            # Product Packet
            product = 1
            for sP in self.subPackets:
                product *= sP.eval()

            return product
        if self.typeID == 2:
            # Minimum Packet
            minimum = float('inf')
            for sP in self.subPackets:
                num = sP.eval()
                if num < minimum:
                    minimum = num

            return minimum
        if self.typeID == 3:
            # Maximum Packet
            maximum = float('-inf')
            for sP in self.subPackets:
                num = sP.eval()
                if num > maximum:
                    maximum = num

            return maximum
        if self.typeID == 4:
            # Literal Packet
            return self.value
        if self.typeID == 5:
            # Greater Than Packet
            return int(self.subPackets[0].eval() > self.subPackets[1].eval())
        if self.typeID == 6:
            # Less Than Packet
            return int(self.subPackets[0].eval() < self.subPackets[1].eval())
        if self.typeID == 7:
            # Equal To Packet
            return int(self.subPackets[0].eval() == self.subPackets[1].eval())
        

def versionSums(packets):
    count = 0
    for p in packets:
        if not p.parsed:
            continue

        count += p.version
        count += versionSums(p.subPackets)
    
    return count


def main(input_path: Optional[Path | str]=None, verbose: bool=False) -> Tuple[Tuple[Any, float]]:
    if not input_path:
        if not (input_path := sys.argv[1] if len(sys.argv) > 1 else None):
            year, day = re.findall(r'\d+', str(__file__))[-2:]
            input_path = Path(Path(__file__).parent.parent.parent, "Inputs", f"{year}_{day}.txt")
    
    with open(input_path, encoding='UTF-8') as f:
        data = [line.strip('\n') for line in f.readlines()]

    with Timer() as p1_time:
        p1 = part1(data)

    if verbose:
        print(f"\nPart 1:\nSum of Version Numbers: {p1}\nRan in {p1_time.elapsed:0.4f} seconds")

    with Timer() as p2_time:
        p2 = part2(data)

    if verbose:
        print(f"\nPart 2:\nEvaulation of Packet: {p2}\nRan in {p2_time.elapsed:0.4f} seconds")

    return [(p1, p1_time.elapsed), (p2, p2_time.elapsed)]


if __name__ == "__main__":
    main(verbose=True)

File: c/16.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <stdbool.h>
#include <ctype.h>
#include <math.h>
#define defaultInput "../../Inputs/2021_16.txt"
#include "../../Modules/input.h"


int bin2int(char *b, int len) {
    int n = 0;

    for (int i = 0; i < len; i++) {
        n *= 2;
        n += b[i] == '1' ? 1 : 0;
    }

    return n;
}


char *parse_packet(char *packet, int *version_sum, long long int *packet_value) {
    int version = bin2int(packet, 3);
    int type_id = bin2int(packet + 3, 3);

    *version_sum += version;

    char *p = packet + 6;

    if (type_id == 4) {
        *packet_value = 0;
        bool continuing = true;

        while (continuing) {
            continuing = p[0] == '1';
            p++;
            *packet_value <<= 4;
            *packet_value += bin2int(p, 4);
            p += 4;
        }
    } else {
        int length_type = p[0] - '0';
        p++;

        long long int *sub_vals = (long long int*)calloc(1, sizeof(long long int));
        int allocated = 1, ix = 0;
        if (length_type == 0) {
            int length = bin2int(p, 15);
            p += 15;

            while (length > 0) {
                while (ix >= allocated) {
                    sub_vals = (long long int*)realloc(sub_vals, 2 * allocated * sizeof(long long int));
                    allocated *= 2;
                }

                int p_len = strlen(p);
                p = parse_packet(p, version_sum, sub_vals + ix);
                ix++;
                length -= p_len - strlen(p);
            }
        } else {
            int num_subpackets = bin2int(p, 11);
            p += 11;

            while (num_subpackets > 0) {
                while (ix >= allocated) {
                    sub_vals = (long long int*)realloc(sub_vals, 2 * allocated * sizeof(long long int));
                    allocated *= 2;
                }

                p = parse_packet(p, version_sum, sub_vals + ix);
                ix++;
                num_subpackets--;
            }
        }

        switch (type_id) {
            case 0:
                *packet_value = 0;
                for (int i = 0; i < ix; i++) {
                    *packet_value += sub_vals[i];
                }
                break;
            case 1:
                *packet_value = 1;
                for (int i = 0; i < ix; i++) {
                    *packet_value *= sub_vals[i];
                }
                break;
            case 2:
                *packet_value = sub_vals[0];
                for (int i = 1; i < ix; i++) {
                    if (sub_vals[i] < *packet_value) {
                        *packet_value = sub_vals[i];
                    }
                }
                break;
            case 3:
                *packet_value = sub_vals[0];
                for (int i = 1; i < ix; i++) {
                    if (sub_vals[i] > *packet_value) {
                        *packet_value = sub_vals[i];
                    }
                }
                break;
            case 5:
                *packet_value = sub_vals[0] > sub_vals[1];
                break;
            case 6:
                *packet_value = sub_vals[0] < sub_vals[1];
                break;
            case 7:
                *packet_value = sub_vals[0] == sub_vals[1];
                break;
        }
    }

    return p;
}

int part1(char *fileName) {
    FILE *fp = fopen(fileName, "r");
    if (!fp) {
        return -1;
    }

    char *input_line = (char*)calloc(2, sizeof(char));
    int ix = 0, max_chars = 1;

    while (!feof(fp)) {
        char c = fgetc(fp);

        if (c != -1) {
            int n = c - '0';
            if (n > 9) {
                n = c - 'A' + 10;
            }

            char *bits = (char*)calloc(5, sizeof(char));
            for (int i = 3; i >= 0; i--) {
                bits[i] = (n % 2) + '0';
                n /= 2;
            }

            while ((ix + 1) * 4 >= max_chars) {
                input_line = (char*)realloc(input_line, (2 * max_chars + 1) * sizeof(char));
                max_chars *= 2;
            }

            strncat(input_line, bits, 4 * sizeof(char));
            ix++;
        }
    }

    int *vSum = (int*)calloc(1, sizeof(int));
    long long int *packetValue = (long long int*)calloc(1, sizeof(long long int));
    parse_packet(input_line, vSum, packetValue);
    return *vSum;
}

long long int part2(char *fileName) {
    FILE *fp = fopen(fileName, "r");
    if (!fp) {
        return -1;
    }

    char *input_line = (char*)calloc(2, sizeof(char));
    int ix = 0, max_chars = 1;

    while (!feof(fp)) {
        char c = fgetc(fp);

        if (c != -1) {
            int n = c - '0';
            if (n > 9) {
                n = c - 'A' + 10;
            }

            char *bits = (char*)calloc(5, sizeof(char));
            for (int i = 3; i >= 0; i--) {
                bits[i] = (n % 2) + '0';
                n /= 2;
            }

            while ((ix + 1) * 4 >= max_chars) {
                input_line = (char*)realloc(input_line, (2 * max_chars + 1) * sizeof(char));
                max_chars *= 2;
            }

            strncat(input_line, bits, 4 * sizeof(char));
            ix++;
        }
    }

    int *vSum = (int*)calloc(1, sizeof(int));
    long long int *packetValue = (long long int*)calloc(1, sizeof(long long int));
    parse_packet(input_line, vSum, packetValue);
    return *packetValue;
}


int main (int argc, char *argv[]) {
    char *inputPath = defaultInput;
    if (argc > 1) {
        inputPath = argv[1];
    }

    clock_t t;
    t = clock(); 
    int p1 = part1(inputPath);
    t = clock() - t; 
    double t_p1 = ((double)t) / CLOCKS_PER_SEC;
    printf("\nPart 1:\nSum of version numbers: %d\nRan in %f seconds\n", p1, t_p1);

    t = clock(); 
    long long int p2 = part2(inputPath);
    t = clock() - t;
    double t_p2 = ((double)t) / CLOCKS_PER_SEC;
    printf("\nPart 2:\nPacket Value: %lld\nRan in %f seconds\n", p2, t_p2);

    return 0;
}